csrc/python_frontend/python_bindings.cpp

```diff
@@ -513,48 +513,132 @@ void initNvFuserPythonBindings(PyObject* module) {
                 !self.completed(),
                 "Attempting to add to a completed definition!");
             Scalar out = self.defineScalar();
-            self.defineRecord(new ScalarRecord<double>(
-                {self.recordingState(out())},
-                serde::RecordType_ScalarInput,
-                std::nullopt,
-                dtype));
+            switch (dtype) {
```

Comment:
I don't think this is needed.  I just think the `ScalarRecord<Type>` serde calls for `define_scalar(None)` matches.

Reply:
The first one matches since `None` will match any `std::optional<T>` regardless of `T`

Reply:
That is true that there is an order problem because of this line:

```
       if (value.has_value()) {                                                               \
          self.defineRecord(new ScalarRecord<CType>(                                           \
              {self.recordingState(out())}, Serde_RType, value, dtype));                       \
```
I think maybe the input version should go back to `ScalarRecord<nullptr_t>` so it is less confusing.🙂

Reply:
Given, the `dtype` and the template type are not necessarily required to be the same.  In fact we really only can see `double` input vals and have to use a `dtype` of `DataType.Float` for input float32s.